tutorial6_hierarchical/demo/main.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * CONSTANTS AND GLOBALS
+ * */
+const width = window.innerWidth * 0.9,
+  height = window.innerHeight * 0.7,
+  margin = { top: 20, bottom: 50, left: 60, right: 40 },
+  tooltipSpecs = { width: 100, height: 50 }
+
+let svg;
+let tooltip;
+
+/**
+ * APPLICATION STATE
+ * */
+let state = {
+  data: null,
+  hover: null,
+};
+
+/**
+ * LOAD DATA
+ * */
+d3.json("../../data/flare.json", d3.autotype).then(data => {
+  state.data = data;
+  init();
+});
+
+/**
+ * INITIALIZING FUNCTION
+ * this will be run *one time* when the data finishes loading in
+ * */
+function init() {
+
+  const container = svg = d3
+    .select("#d3-container")
+    .style("position", "relative")
+
+  tooltip = container 
+    .append("div")
+    .attr("width", 100)
+    .attr("height", 100) 
+    .style("position", "absolute")
+    .style("background-color", "white")
+
+  svg = container
+    .append("svg")
+    .attr("width", width)
+    .attr("height", height);
+
+  // groups the data by genre, type and rating
+  // make hierarchy
+  const root = d3
+    .hierarchy(state.data) // children accessor
+    .count() // sets the 'value' of each level
+    .sort((a, b) => b.value - a.value);
+
+  // make treemap layout generator
+  const tree = d3
+    .treemap()
+    .size([width, height])
+    .padding(1)
+    .round(true);
+
+  // call our generator on our root hierarchy node
+  tree(root); // creates our coordinates and dimensions based on the heirarchy and tiling algorithm
+
+  console.log(root);
+
+  // create g for each leaf
+  const leaf = svg
+    .selectAll("g")
+    .data(root.leaves())
+    .join("g")
+    .attr("transform", d => `translate(${d.x0},${d.y0})`);
+
+  // create title for each node
+  leaf.append("title").text(
+    d =>
+      `${d
+        .ancestors()
+        .reverse()
+        .map(d => d.data.name)
+        .join("/")}\n${d.value}`
+  );
```

Comment:
we can probably remove this for the title if we have the tooltip -- though I will say that it is nice to see the whole path down to the selected node, so it might be worth bringing this logic into the tooltip name.

Reply:
I think this is helpful while we discuss in class but it actually gets in the way of the tooltip in the current implementation, so I put it in the tooltip